File: frappe-bench/apps/myschools/myschools/public/js/setup_wizard.js
// MY Schools — setup wizard slide.
//
// Registered via the `setup_wizard_requires` hook in myschools/hooks.py.
// Adds one extra slide *after* ERPNext's stock slides (organization /
// user-details / region). All fields are optional — operators who don't
// want to seed the franchise tree at install time can leave them blank
// and create Cluster / Branch / Campus through the normal forms later.
//
// The values collected here are POSTed alongside ERPNext's args to
// `frappe.desk.page.setup_wizard.setup_wizard.setup_complete` and are
// read in Python by `myschools.scripts.setup_wizard.create_first_franchise_tree`.

frappe.provide("mys.setup");

frappe.setup.on("before_load", function () {
	// Skip if this app's setup already completed on a prior wizard run.
	if (
		frappe.boot.setup_wizard_completed_apps?.length &&
		frappe.boot.setup_wizard_completed_apps.includes("myschools")
	) {
		return;
	}
	mys.setup.slides_settings.map(frappe.setup.add_slide);
});

mys.setup.slides_settings = [
	{
		name: "mys_franchise",
		title: __("Seed your first Cluster, Branch and Campus"),
		// Optional slide — operators can skip every field.
		icon: "fa fa-sitemap",
		fields: [
			{ fieldtype: "Section Break", label: __("First Cluster (optional)") },
			{
				fieldname: "mys_cluster_code",
				label: __("Cluster Code"),
				fieldtype: "Data",
				description: __("Short code, e.g. CLR-NORTH. Leave blank to skip."),
			},
			{ fieldtype: "Column Break" },
			{
				fieldname: "mys_cluster_name",
				label: __("Cluster Name"),
				fieldtype: "Data",
			},
			{ fieldtype: "Column Break" },
			{
				fieldname: "mys_cluster_region",
				label: __("Region"),
				fieldtype: "Data",
				description: __("Optional descriptor, e.g. 'Northern Punjab'."),
			},

			{ fieldtype: "Section Break", label: __("First Branch (optional)") },
			{
				fieldname: "mys_branch_code",
				label: __("Branch Code"),
				fieldtype: "Data",
				description: __("Short code, e.g. BR-001. Required to create a Branch."),
			},
			{ fieldtype: "Column Break" },
			{
				fieldname: "mys_branch_name",
				label: __("Branch Name"),
				fieldtype: "Data",
			},

			{ fieldtype: "Section Break", label: __("First Campus (optional)") },
			{
				fieldname: "mys_campus_type",
				label: __("Campus Type"),
				fieldtype: "Select",
				options: "\nKids\nJunior\nSenior",
				description: __("Pick a tier. Will be created under the Branch above."),
			},
		],

		validate: function () {
			// Anything goes — the Python stage skips rows whose required
			// fields are blank, so an entirely blank slide is valid (the
			// user simply doesn't seed any data at install time).
			//
			// Light validation: if any Branch field is filled, both must be,
			// otherwise the row would silently be dropped on the backend.
			const v = this.values;
			const branch_partial =
				(v.mys_branch_code && !v.mys_branch_name) ||
				(v.mys_branch_name && !v.mys_branch_code);
			if (branch_partial) {
				frappe.msgprint(__("Enter both Branch Code and Branch Name, or leave both blank."));
				return false;
			}
			const cluster_partial =
				(v.mys_cluster_code && !v.mys_cluster_name) ||
				(v.mys_cluster_name && !v.mys_cluster_code);
			if (cluster_partial) {
				frappe.msgprint(__("Enter both Cluster Code and Cluster Name, or leave both blank."));
				return false;
			}
			// Branch needs a Cluster — either from this slide or pre-existing.
			if (v.mys_branch_code && !v.mys_cluster_code) {
				frappe.msgprint(__("A Branch needs a Cluster. Fill in the Cluster fields above."));
				return false;
			}
			// Campus needs a Branch.
			if (v.mys_campus_type && !v.mys_branch_code) {
				frappe.msgprint(__("A Campus needs a Branch. Fill in the Branch fields above."));
				return false;
			}
			return true;
		},
	},
];

File: frappe-bench/apps/myschools/myschools/scripts/setup_wizard.py
"""Custom setup-wizard stages for MY Schools.

Registered via the `setup_wizard_stages` hook. Runs *after* ERPNext's
stages (presets → company → defaults → wrap-up), so by the time
`create_first_franchise_tree` fires, the Company exists and the user
has been logged in. We use that Company as the default link for the
first Branch.

The slide that collects the field values lives in
`public/js/setup_wizard.js` and is registered via the
`setup_wizard_requires` hook.

Every field is optional — operators who skip the slide get a vanilla
ERPNext install plus our app, and create their first Cluster / Branch /
Campus through the normal forms later.
"""

import frappe
from frappe import _


def get_setup_stages(args=None):
	return [
		{
			"status": _("Creating first franchise tree"),
			"fail_msg": _("Failed to create first franchise tree"),
			"tasks": [
				{
					"fn": create_first_franchise_tree,
					"args": args,
					"fail_msg": _("Failed to create first franchise tree"),
				}
			],
		}
	]


def create_first_franchise_tree(args):
	"""Optionally create one Cluster + Branch + Campus from wizard input.

	Each row is created only if its required fields are present. Skips
	silently if a record with the same primary key already exists, so
	re-running the wizard (or running this from ci_bootstrap) is
	idempotent.
	"""
	args = frappe._dict(args or {})
	cluster = _maybe_create_cluster(args)
	branch = _maybe_create_branch(args, cluster_code=cluster)
	_maybe_create_campus(args, branch_code=branch)
	frappe.db.commit()


def _maybe_create_cluster(args):
	code = (args.get("mys_cluster_code") or "").strip()
	name = (args.get("mys_cluster_name") or "").strip()
	if not (code and name):
		return None
	if frappe.db.exists("MYS Cluster", code):
		return code
	frappe.get_doc(
		{
			"doctype": "MYS Cluster",
			"cluster_code": code,
			"cluster_name": name,
			"region": args.get("mys_cluster_region") or "",
			"is_active": 1,
		}
	).insert(ignore_permissions=True)
	return code


def _maybe_create_branch(args, cluster_code):
	code = (args.get("mys_branch_code") or "").strip()
	name = (args.get("mys_branch_name") or "").strip()
	if not (code and name and cluster_code):
		return None
	if frappe.db.exists("MYS Branch", code):
		return code
	company = frappe.db.get_value("Company", {}, "name")
	frappe.get_doc(
		{
			"doctype": "MYS Branch",
			"branch_code": code,
			"branch_name": name,
			"cluster": cluster_code,
			"company": company,
			"is_active": 1,
		}
	).insert(ignore_permissions=True)
	return code


def _maybe_create_campus(args, branch_code):
	campus_type = (args.get("mys_campus_type") or "").strip()
	if not (branch_code and campus_type):
		return None
	# MYS Campus autoname is `format:{branch}-{campus_type}`.
	name = f"{branch_code}-{campus_type}"
	if frappe.db.exists("MYS Campus", name):
		return name
	frappe.get_doc(
		{
			"doctype": "MYS Campus",
			"branch": branch_code,
			"campus_type": campus_type,
			"is_active": 1,
		}
	).insert(ignore_permissions=True)
	return name
